File: ios/Plugin/FFmpegUtil.swift
//
//  FFmpegUtil.swift
//  Plugin
//
//  Created by Tenda-xue on 2022/3/14.
//  Copyright © 2022 Max Lynch. All rights reserved.
//

import Foundation
import mobileffmpeg
import Capacitor


@objc public class FFmpegUtil: NSObject, ExecuteDelegate {
    
    var _call: CAPPluginCall?;
    var _outPutFile: String = "";
    
    /**
     视频截图
     */
    @objc public func ffScreenshot(videoFile: String, atTime: Float , imageW: Float , imageH: Float, call: CAPPluginCall) {
        _call = call;
        let cachePath = NSHomeDirectory() + "/Library/Caches/"
        let  now =  NSDate ();
        let outPutFile: String =  String(format: "%@%d_%f_ff_thumbnail.jpg", cachePath,Int(now.timeIntervalSince1970),atTime);
        let cmd = String(format: " -ss %f -i %@ -f image2 -s %dx%d  -vframes 1 %@", atTime, videoFile, Int(imageW), Int(imageH), outPutFile);
        MobileFFmpeg.executeAsync(cmd, withCallback: self);
        _outPutFile = outPutFile;
    }
    
    /**
     视频转码
     */
    @objc public func ffTransCode(videoFile: String, videoBitrate: String , call: CAPPluginCall ){
        _call = call;
        let cachePath = NSHomeDirectory() + "/Library/Caches/"
        let  now =  NSDate ();
        let outPutFile: String =  String(format: "%@%d_ff_trans.mp4", cachePath,Int(now.timeIntervalSince1970));
        let cmd = String(format: " -i %@ -b:v %@ -y %@", videoFile,videoBitrate, outPutFile);
        MobileFFmpeg.executeAsync(cmd, withCallback: self);
        _outPutFile = outPutFile;
    }
 
    /**
     ffmpeg回调
     */
    public func executeCallback(_ executionId: Int, _ returnCode: Int32) {
        _call!.resolve([
            "resultCode": returnCode,
            "outPutFile": _outPutFile
        ])
    }
    

}

File: android/src/main/java/com/tenda/aio/FFmpegUtil.java
package com.tenda.aio;

import android.content.Context;

import com.arthenica.mobileffmpeg.ExecuteCallback;
import com.arthenica.mobileffmpeg.FFmpeg;
import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;

public class FFmpegUtil {

    private Context context;

    public void setContext(Context context) {
        this.context = context;
    }

    /**
     * 视频截图
     * @param videoFile
     * @param atTime
     * @param call
     */
    public void ffScreenshot(String videoFile, Float atTime, Float imageW,Float imageH,PluginCall call){
        String cachePath = this.context.getCacheDir().getPath();
        String outPutFile = cachePath + System.currentTimeMillis() +"_ff_thumbnail.jpg";
        String cmd = String.format(" -ss %f -i %s -f image2 -s %dx%d -vframes 1 %s",atTime,videoFile,Math.round(imageW),Math.round(imageH),outPutFile);
        FFmpeg.executeAsync(cmd, new ExecuteCallback() {
           @Override
           public void apply(long executionId, int returnCode) {
               JSObject ret = new JSObject();
               ret.put("resultCode", returnCode);
               ret.put("outPutFile", outPutFile);
               call.resolve(ret);
           }
        });
    }

    /**
     * 视频转码
     * @param videoFile
     * @param videoBitrate
     * @param call
     */
    public void ffTransCode(String videoFile, String videoBitrate, PluginCall call){
        String cachePath = this.context.getCacheDir().getPath();
        String outPutFile = cachePath + System.currentTimeMillis() +"_ff_trans.mp4";
        String cmd = String.format(" -i %s -b:v %s -y %s",videoFile,videoBitrate,outPutFile);
        FFmpeg.executeAsync(cmd, new ExecuteCallback() {
            @Override
            public void apply(long executionId, int returnCode) {
                JSObject ret = new JSObject();
                ret.put("resultCode", returnCode);
                ret.put("outPutFile", outPutFile);
                call.resolve(ret);
            }
        });
    }
}
